### ISSUE TITLE
Nlog config provider abstraction for a CMS -- v5 candidate

### ISSUE THREAD

**Issue by ericnewton76** [NLog/NLog#3055]
Running into a core architectural problem with NLog, trying to get the source to compile and run on a CMS with very limited file access.  This CMS is C# based, and webfarm enabled, so writing/loading from local file system is undeterministic due to load balancing/etc.  All System.IO calls are denied. 

Basically I have to call the CMS's asset loading mechanisms and be able to provide file system alternative access instead of IF DEFINED pragmas all over.  (LogFactory is the worst... lots of leaky file access in there, plus calling out to XmlConfiguration and various)

As I look through the code base, there's a lot (a whole lot) of attempts at dealing with file system idiosyncrisis between Andriod iOS windows UWP and various others...

 - Why do we need it?
 Create legit abstraction for config file loading for various platforms.

My course of action would be ripping out every bit of IFile/FileWrapper/etc from LogFactory and simply using an NLogConfigurationProvider type of mechanism.   The real "default" might need to come from NLog.Platform. specific package and the built in default simply runs a INFO or higher rule to a fifo linked list with 1000 messages max stored in memory with a GetMessages() specific only to it somehow. (ie not on an interface but maybe it should be)

I would then need to create Targets for my CMS, and a custom CMSConfigProvider that could load the configuration either explicitly or via first log message by making CMS API calls.  For me, there's a race condition that I can't do anything about... Log messages might be invoked before I can supply that CMSConfigProvider to the LogFactory but as of now, there's not much I can do about it... the CMS doesnt supply an "OnStartup" type of facility (yet).  Thats okay... something is better than nothing! (Or in this case NLogging is better than the builtin CMS logging... basically like old school ASP Response.Write... bleh!)

I feel like v5 is the right time for a major overhaul of that aspect of the NLog platform.  What do you guys think?

**Comment by ericnewton76** [comment]
Btw, are you running a branch with v5 or just tags?  I cant seem to tell right now.

Seems like you still need a branch for v4.6 for any "hotfixes"/oopsies

**Comment by ericnewton76** [comment]
Whats interesting, in LogFactory, pretty much every `#if !SILVERLIGHT` i was adding a `#if CMS_IGNORE && !SILVERLIGHT` and that started making things work but theres still problems with that approach

**Comment by snakefoot** [comment]
Interesting exercise with using NLog in a environment without file-access. Would make sense to extract all  nlog-extension-assembly-loading-with-nlog-config-file-loading-and-monitoring-for-reload from LogFactory and XmlLoggingConfiguration into an isolated class that implements a common interface.

The question is whether there is some way for you to initialize NLog, so it will use a different interface-implementation (That doesn't do file-access) instead of the default one?

Is the CMS running on NetCore or NetFramework?

**Comment by snakefoot** [comment]
@ericnewton76 Would it be possible to share how many places you needed to add `CMS_IGNORE` for NLog to work for your CMS? (With PullRequest or with link to forked repository with your changes committed)

**Comment by ericnewton76** [comment]
> Is the CMS running on NetCore or NetFramework?

@snakefoot CMS is running NetFramework 4.6.2 with csc compiler that doesnt support `=>` shorthand functions.  So after you read that last sentence, yes, I'm having to also adapt the NLog source code to run within the CMS... it has its own build system, that you load all the source code into, plus some code for the templates that govern how assets are handled.  The CMS' build system was built probably to "control" the code that goes in... imagine not having the ability to reference 3rd party dlls... (theres a good way to prevent security and file access issues... lol)  

And since I have to load source files in, and dont wanna maintain 50+ files, I've also built a source packer, that can take all the source code files and pack them into a single file.  Using Roslyn, It involves consolidating global usings and a simple extraction of whatever's defined within the `namespace` keywords.  Currently its known as cppacker but I'm already about to fold that into a bigger project called cppm.

I also need to explain how I'm transitioning NLog into the CMS.  I will follow up with another post about that.  Actually, go see https://github.com/cpcontrib/cpcontrib.logging/issues/1 it'll help explain a lot

> The question is whether there is some way for you to initialize NLog, so it will use a different interface-implementation (That doesn't do file-access) instead of the default one? Possible for you to stay away from LogManager and instead maintain singleton-lifetime of a self-created LogFactory-instance ? (Constructed to use a different interface-implementation)

> Looks like the changes needed for LogFactory are rather isolated, so could probably be implemented as part of NLog 4.6 (Allowing one to override the config-loader). But by default using config-loader that scans and autoloads NLog.config files.

The changes do feel isolated, yet numerous just because of the scattershot of locations in LogFactory.cs of the file access routines.  There are some examples of file access spidering out into a few other config related stuff.  And LogFactory directly calling XmlConfigurator itself. 

Since there's an interest I will start doing more yanking and decoupling in a fork of NLog instead of scalpel patching.  The big issue is the file based access within LogFactory that conceptually moves to a ConfigurationProvider.  

I'm getting a feel for the entirety of the code base because I have been importing each file one-by-one, rewriting code that compiles for csc v7+ for csc <v7.

**Comment by ericnewton76** [comment]
> Looks like the changes needed for LogFactory are rather isolated, so could probably be implemented as part of NLog 4.6 (Allowing one to override the config-loader). But by default using config-loader that scans and autoloads NLog.config files.

What does the "default" loader do?  I know it looks for Nlog.config in the assembly codebase area plus a few other tries, but when there's no NLog.config at all, what happens?

I think for the CMS, we wont ever get a hit on a file in the filesystem for config... it just wont be there.  On the other hand, if the CMS itself adopts NLog (it currently interally uses log4net) the config loader would then conflict, so there's gonna have to be a way to control the loading behavior.

My first thought would be an appsetting key, where the default NLog configuration looking behavior is preserved.  Something like: `<appSetting key="NLog.ConfigurationProviderType" value="CMSConfigurationProvider" />`    I think appsetting key is best because its the most basic way to get the change in before the appdomain even loads.

**Comment by ericnewton76** [comment]
I created a fork, and a new branch:
https://github.com/ericnewton76/nlog/tree/feat/configuration-provider

**Comment by snakefoot** [comment]
> with csc compiler that doesnt support => shorthand functions. So after you read that last sentence, yes, I'm having to also adapt the NLog source code to run within the CMS

Funny decision with having to compile everything to use it in the CMS. 

> The CMS' build system was built probably to "control" the code that goes in... imagine not having the ability to reference 3rd party dlls

What happens if you just call [Assembly.Load](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly.load) of `NLog.dll` and then perform creation of the NLog LogFactory using [Activator.CreateInstance](https://docs.microsoft.com/en-us/dotnet/api/system.activator.createinstance) would it not be possible to load NLog without having to convert/recompile all code? (Maybe with help from [LibLog](https://github.com/damianh/LibLog))

>What does the "default" loader do? I know it looks for Nlog.config in the assembly codebase area plus a few other tries, but when there's no NLog.config at all, what happens?

The current behavior in NLog is to scan different locations for a `NLog.config`-file, and if none found it just  continues without any logging-rules or logging-targets. Until a logging-configuration is manually loaded.

> The CMS currently internally uses log4net

If it can run log4net then it should have no issues with NLog, unless it is a custom tailored implementation of log4net. It is possible to pipe the output from log4net to NLog. Either by a custom [NLog-appender](https://www.yqz.dk/archive/forward-sitecore-logs-to-nlog/) or by using [Log4net TraceAppender](https://logging.apache.org/log4net/release/sdk/html/T_log4net_Appender_TraceAppender.htm) together with [NLogTraceListener](https://github.com/NLog/NLog/wiki/Send-System-Diagnostic-(Trace)-to-NLog) (All should be configurable through app.config, both log4net-config and NLog-config)

**Comment by snakefoot** [comment]
> There's a lot of inaccurate testing going on with LogFactory... LogFactoryTests are testing LogManager... and yes, LogManager sould be tested but LogFactory is what LogManager is wrapping its static calls to.

I think you see the result of a baby with many fathers and many mothers, not all parents sees the finer details. Nit-picking and volunteer work in spare time doesn't always go hand in hand. It is a constant effort to keep code clean and simple, especially for a codebase as old as NLog.

**Comment by ericnewton76** [comment]
Oh yeah I get that... lol.   The packer is relatively simple concept and still works for the most part... its only problem is getting fed code files that dont compile with the older compilers.  I might extend the packer since its using Roslyn to rewrite statements like that... not sure... maybe over some beers and a smoke or something.

Assembly Loading is pretty much a no-go as well, although I'm tempted to force their hand by invoking it.  Only problem then is when the template code is compiled, I cant add references to the NLog assembly so the compile of that code fails.  Its quite a unique limiting situation.

LibLog might be the solution, where I can cheat a bit, load the NLog assembly from the Asset repository into memory and then LibLog hooks up to it.

Although I'm still eager to decouple LogFactory from files... it only makes sense to try to do this with you guys' blessing of course... I believe it warrants a v5 requirement just based on the breadth of changes I've already made today to try to do that decoupling... Theres a lot of interface changes that remove so many roundabout ways that have been baked in and added on over several years.  My hope is that this doesnt stay as a separate fork of NLog for very long but it may end up being a good thing for a while.

**Comment by ericnewton76** [comment]
I think one thing to consider in all this is what public interface do you (we) actually wanna support going forward?

I would guess:
- XML file structure
- ILogger / Logger interfaces
- LogManager
- XmlConfigurator.ConfigureAndWatch(file)

and code like
```c#
            var config = new LoggingConfiguration();
            var consoleTarget = new ColoredConsoleTarget();
            config.AddTarget("console", consoleTarget);
            config.LoggingRules.Add(new LoggingRule("*", LogLevel.Debug, consoleTarget));
            LogManager.Configuration = config;
```

To do that, I wonder if theres an attribute like `[Solid]` or `[ForAllTime]` or `[NeverGonnaChange]` or something similar

**Comment by snakefoot** [comment]
The ILogger / Logger-interfaces are very close to carved into stone. Everything else is pretty much up in the air to break when it makes sense with major version bump.

Usually the road is to mark things as obsolete first, and then remove them when it makes sense.

**Comment by ericnewton76** [comment]
ILogger/Loggers carved in stone... totally agree 100%.  

For v4.6 for sure, how about for v5 ?  Can we just draw a line in the sand and start disappearing public members besides that list above?  Does it make sense to draw up a roadmap of public members for v5 that will just go away... (lol)?  Maybe we create a file in the wiki area to outline that thought process and link to it from the README.md?  (Thinking out loud as I usually do)

**Comment by ericnewton76** [comment]
Btw, thanks for that LibLog suggestion... oh man that just might make it work sooner for me...  :smile:

**Comment by snakefoot** [comment]
Created #3058 that moves the file-access code out of LogFactory.

**Comment by snakefoot** [comment]
@ericnewton76 Guess this issue can be closed (Together with #3056) since your quest have stopped :)

**Comment by ericnewton76** [comment]
I do dislike never-ending items.  I'll request reopen if the need arises or if I can bring some closure to the situation.  

Rolf did a lot of work reworking elements of it, in a much different approach then I took, so my PR changes are redundant.
